File: frontend/lib/core/money.dart
/// Las cuentas de dinero, en un solo lugar.
///
/// Regla de todo el proyecto: el dinero NUNCA pasa por double. En Dart,
/// 0.1 + 0.2 da 0.30000000000000004, y ese centavo fantasma termina en la
/// cuenta de alguien.
///
/// La solucion es la que usa cualquier sistema de facturacion: se trabaja en
/// CENTAVOS (numeros enteros) y se convierte a texto solo para mostrar. Asi
/// 150 + 100 es 250, exacto, sin importar cuantas veces se sume.
///
/// El backend hace exactamente lo mismo con bcmath (ver App\Support\Money).
/// Los dos lados tienen que redondear igual, porque es lo unico que impide que
/// la pantalla muestre un total y el servidor cobre otro.
class Money {
  /// "1.50" -> 150.
  ///
  /// Devuelve 0 si el texto no es un numero valido. La app nunca deberia
  /// mandar basura, pero un carrito que revienta por eso es peor que un
  /// carrito que muestra 0.
  static int toCents(String amount) {
    final String clean = amount.trim();

    if (clean.isEmpty) {
      return 0;
    }

    final List<String> parts = clean.split('.');

    final int? whole = int.tryParse(parts[0]);
    if (whole == null) {
      return 0;
    }

    // OJO: "1.5" son 50 centavos, no 5. Hay que rellenar a la derecha antes
    // de cortar, o el precio se cae a la decima parte.
    String decimals = parts.length > 1 ? parts[1] : '';
    decimals = decimals.padRight(2, '0').substring(0, 2);

    final int? cents = int.tryParse(decimals);
    if (cents == null) {
      return 0;
    }

    return whole * 100 + cents;
  }

  /// 150 -> "1.50".
  static String fromCents(int cents) {
    final int whole = cents ~/ 100;
    final int rest = cents % 100;

    return '$whole.${rest.toString().padLeft(2, '0')}';
  }

  /// Suma montos en texto: `add(['1.50', '2.00'])` -> "3.50".
  static String add(Iterable<String> amounts) {
    int total = 0;

    for (final String amount in amounts) {
      total += toCents(amount);
    }

    return fromCents(total);
  }

  /// Precio por cantidad: `multiply('1.25', 3)` -> "3.75".
  static String multiply(String amount, int quantity) {
    return fromCents(toCents(amount) * quantity);
  }
}

File: backend/app/Support/Money.php
<?php

namespace App\Support;

/**
 * Las cuentas de dinero, en un solo lugar.
 *
 * Regla de todo el proyecto: el dinero NUNCA pasa por float. Un float de PHP
 * no puede representar 0.10 exacto, asi que 1.10 + 2.20 da 3.3000000000000003
 * y ese centavo fantasma aparece en la cuenta de alguien.
 *
 * Por eso los montos viajan como TEXTO ("1.50") desde la base hasta Flutter,
 * y las sumas se hacen con bcmath, que trabaja en decimal exacto.
 *
 * Esta clase existe para que la regla se aplique igual en todos lados: si
 * cada controlador sumara a su manera, alcanzaria con que uno solo use '+' para
 * que los numeros dejen de cuadrar.
 */
class Money
{
    /**
     * Cuantos decimales tiene la moneda. Todo el proyecto usa 2.
     */
    private const SCALE = 2;

    /**
     * Suma dos o mas montos.
     */
    public static function add(string ...$amounts): string
    {
        $total = '0.00';

        foreach ($amounts as $amount) {
            $total = bcadd($total, $amount, self::SCALE);
        }

        return $total;
    }

    /**
     * Multiplica un monto por una cantidad entera (precio x cantidad).
     */
    public static function multiply(string $amount, int $quantity): string
    {
        return bcmul($amount, (string) $quantity, self::SCALE);
    }

    /**
     * Resta dos montos.
     */
    public static function subtract(string $amount, string $other): string
    {
        return bcsub($amount, $other, self::SCALE);
    }

    /**
     * ¿El monto es mayor que cero?
     */
    public static function isPositive(string $amount): bool
    {
        return bccomp($amount, '0', self::SCALE) === 1;
    }
}
